[AST][Sema] Typed throws implementation

minuscorp opened the pull request:
<!-- What's in this pull request? -->
This PR adds a basic implementation for adding a type after the `throws` keyword as discussed in the following thread: https://forums.swift.org/t/typed-throws/39660

**This is a draft**

Todo:

- [ ] Add tests
- [x] Update mangling (if needed)
- [X] Parsing and parse tests.
- [ ] Type-checking throw type.

minuscorp committed:
Add basic AST representation of the throws type representation

minuscorp commented:
Developers are welcome in order to get this implementation going forward (as it is needed for a review process).

dfsweeney commented:
I think some of the problems building the stdlib are from a lot of diagnostic noise when trying to parse a type and failing. That's straightforward to fix--you can do something like 

```
    if (!peekToken().isKeyword()) {
      BacktrackingScope backtrackingScope(*this);
      DiagnosticSuppression diagnosticSuppression(Diags);

       ParserResult<TypeRepr> result = parseType();
       if (result.isNonNull()) {
         throwsType = result.get();
       }
```

when the DiagnosticSuppression goes out of scope the diagnostics go back on. I added that locally and got some improvement.

The one that's generating a lot of noise compiling the stdlib is 

```
   if (Tok.is(tok::arrow)) {
-    ParserResult<TypeRepr> ty =
-      parseTypeSimpleOrComposition(MessageID, HandleCodeCompletion);
-    if (ty.isNull())
-      return ty;
-    auto tyR = ty.get();
-    auto status = ParserStatus(ty);
     // Handle type-function if we have an arrow.
     SourceLoc arrowLoc = consumeToken();
```

If you take that out, then stdlib compiles, but obviously that defeats the purpose of the whole thing. 

I want to make sure I understand that one though. If the current token is an arrow, haven't we moved past anything like `throws SomeError` ? In the old syntax it's like ` _ body: (UnsafeBufferPointer<C.Element>) throws -> R`, in the new syntax it's ` _ body: (UnsafeBufferPointer<C.Element>) throws SomeError-> R` right? If the current token is an arrow then we are past the throw type. Or am I misunderstanding? I might be missing the context.

minuscorp commented:
The new syntax is as you said `body: (UnsafeBufferPointer<C.Element>) throws SomeError-> R` that's why I try to get the type using `parseTypeOrComposition` before the arrow, and after the potential `throws` being found, and just if that type exists, that can of course not be there because it is optional.

minuscorp committed:
Green build

theblixguy commented:
You can call `canParseType()` followed by `parseType()` if the former succeeds.

minuscorp committed:
Backtrack to find the throws keyword

minuscorp committed:
Check with canParseType()

minuscorp committed:
Added  syntax tree nodes

minuscorp commented:
As it was an optional token, I chose to get the type if was present or null in other case scenario. With `canParseType()` I'll always get the `TypeRepr` but I have to initialize it to `nullptr` for the non existent case. Is it worth?

minuscorp committed:
Syntax update

theblixguy commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -849,6 +851,19 @@ void Parser::parseAsyncThrows(
+    if (!peekToken().isKeyword()) {
+      BacktrackingScope backtrackingScope(*this);
+      if (peekToken().is(tok::kw_throws)) {
+        ASTContext &Ctx = SF.getASTContext();
+        DiagnosticSuppression SuppressedDiags(Ctx.Diags);
+        backtrackingScope.cancelBacktrack();
```
Why is this needed?

minuscorp replied:
I use the backtrack for searching of my previous word was a keyword and if it was the throws keyword, but maybe there's some other way to do it.

theblixguy replied:
I think what you should do is something like:

```c++
bool hasType = false;
{
  BacktrackingScope backtrack(*this);
  hasType = canParseType();
}

if (hasType) {
  ParserResult<TypeRepr> result = parseType();
  throwsType = result.getPtrOrNull();
}
```

minuscorp replied:
That is good for consuming the type token! Thank you! But I'm unable to check that the previous token was `tok::throws`

dfsweeney commented:
ok, `errors.swift` fails with:

```
/Users/dfs/swift-source/swift/test/Parse/errors.swift:47:8: error: unexpected error produced: expected '{' in body of function declaration
  func bar() throws SomeError {}
       ^
/Users/dfs/swift-source/swift/test/Parse/errors.swift:47:20: error: unexpected error produced: consecutive statements on a line must be separated by ';'
  func bar() throws SomeError {}
                   ^
/Users/dfs/swift-source/swift/test/Parse/errors.swift:47:31: error: unexpected error produced: argument passed to call that takes no arguments
  func bar() throws SomeError {}
                              ^
```

working through that in `lldb` it looks like 

```
if (!peekToken().isKeyword()) {
   	      BacktrackingScope backtrackingScope(*this);
 	      if (peekToken().is(tok::kw_throws)) {
```
is skipping the important part--we already consumed the 'throws' and the active token might is the throw type. We should try to parse the type immediately, and not look for `throws` I made a change and am building locally.

minuscorp commented:
If we don't check for `throws` we're gonna end up with a messed order of elements, don't we?

dfsweeney commented:
> If we don't check for `throws` we're gonna end up with a messed order of elements, don't we?

I think the order of elements should be good--we don't need to check for `throws` because we already found it, is what it looks like. I could still be missing it. 

The stdlib won't build after my change so it needs something else. I'm interpreting what's happening after ParsePattern.cpp:840 as:

1. We found 'throws' (or 'try' or 'throw') 
1. then at 853 we consume the token and set throwsLoc with the token
1. So we are pointing at whatever is after the `throws`, either a type, or something else (`->` or `{`, probably, but there are other possibilities)
1. we call `canParseType` and if it succeeds we found a type and should then really parse it and put it in `throwsType`. 

I think where the problem is after my change is that `canParseType` may be interpreting some things that are not actually types but that are legal Swift as if they were types. I have to look in to that more.

minuscorp committed:
Parse Type after seeing and consuming throws

minuscorp commented:
I got it working

minuscorp commented:
Could we make a quick testrun and a source compatibility on the CI? (tests are passing locally)

dfsweeney commented:
> Altough I cannot make stdlib to compile, lots of errors (unrelated), to the changes being made (?)

What are the errors like? I am having problems compiling 

```
  mutating func removeAll(
    where shouldBeRemoved: (Element) throws -> Bool) rethrows

  // FIXME: Associated type inference requires these.
  @_borrowed
``` 
in `RangeReplaceableCollection.swift`, I think because the `rethrows` is using some of the new logic, although I'm not sure.

minuscorp commented:
Things like:
```
/Users/minuscorp/Documents/swift-source/swift/stdlib/public/core/UnsafeBufferPointer.swift.gyb:197:14: error: cannot convert value of type '()' to expected condition type 'Bool'
    if n > 0 ? l >= 0 && l < n : l <= 0 && n < l {
       ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

minuscorp commented:
Now I'm only getting:
```
error: consecutive declarations on a line must be separated by ';'
  mutating func encode(_ value: Float, forKey key: Key) throws
          ^
          ;
```
Only `throws` related errors in the stdlib

dfsweeney commented:
is that in `KeyedEncodingContainerProtocol`? The keyword before `mutating` is `throws` (the comments don't count) and it's a protocol so `throws` terminates the previous method signature (or whatever it's called in a protocol.) But it's reading ahead into the `mutating` and getting confused.

minuscorp commented:
I'm getting a hard time with Deserialization, I'll push what I have but I does not compile, I cannot find a way to get from the deserialized func to a `TypeRepr`

minuscorp committed:
Progress... I think

minuscorp committed:
Get type or null

dfsweeney committed:
Added parsing of 'throws (Type)' syntax. (#2)

* Added parsing of 'throws (Type)' syntax.

New diagnostic if the parsing fails. Logic to parse an optional
parenthesized type identifier after the 'throws' keyword.

This will not compile the standard library right now because it cannot
parse the pattern

  `func f(_ body: (T) throws -> U) rethrows -> U {return try body(t)}`

That pattern is really common in the standard library for things like
`withUnsafeBufferPointer`. I'm not sure what the problem is right now there.

* Added Parser::parseThrowsType.

Edited Parse/errors.swift test to include
some throws (SomeError) entries. Parse/errors.swift passes in this form.
Stdlib compiles. Parse/async.swift, Syntax/Parser/tree.swift and
incrParse/incrementalTransfer.swift fail right now. Also I think I need to set
a SourceLoc for the throwsType explicitly.

* [Parse] Parse type after throws correctly in function declarations

* [Parse] Parse type after throws correctly in function types

* Improve documentation of new parsing functions

Co-authored-by: Josef Zoller <josef@walterzollerpiano.com>

minuscorp committed:
Syntax tree

minuscorp committed:
Updated tree declaration

minuscorp committed:
Parser and AST Additions

minuscorp committed:
Fix typesyntaxtests

minuscorp committed:
Fix errors tests, broke async-syntax ones (compile crash)

minuscorp committed:
Update lib/Parse/ParsePattern.cpp

minuscorp committed:
Fix review comments

Zollerboy1 committed:
Correct the syntax tree for function declarations and types

Zollerboy1 committed:
Fix FunctionTypeRepr constructor and some code style issues

minuscorp committed:
Fix round_trip_parse_gen tests

minuscorp committed:
Fixes for round_trip_tests and syntax tree

minuscorp committed:
Fix incremental test

Zollerboy1 committed:
Fix Syntax/Parser/tree.swift test

minuscorp committed:
Merge remote-tracking branch 'origin/master' into typed-throws-dev

minuscorp committed:
Merge branch 'master' into typed-throws-dev

minuscorp committed:
Merge pull request #11 from minuscorp/typed-throws-dev

Update typed throws implementation

Zollerboy1 commented:
This is all parsing done.

minuscorp committed:
Merge branch 'master' into typed-throws-1

minuscorp commented:
@swift-ci please smoke test

minuscorp commented:
cc @theblixguy ?

theblixguy commented:
You need commit access to trigger CI. I’ll do it for you.

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci please test source compatibility

minuscorp commented:
and how do I get with those permissions? as we're working on a proposal implementation and will need it sometimes, do I ask you?

minuscorp commented:
Jenkins died before the smoke test even started on OS X

theblixguy commented:
You can run the tests locally. If you have any problems with that, then I can run CI for you if you ping me.

theblixguy commented:
@swift-ci please smoke test macOS

theblixguy commented:
Looks like there’s an infrastructure problem preventing smoke tests on macOS. Let’s wait for it to be resolved.

minuscorp commented:
Yeah we always test locally before updating the PR but as the CI is the source of truth we want to make sure we didn't screw up anything locally.

minuscorp commented:
Two questions @theblixguy, why the compat may fail in debug and not in release? which the difference? on the other hand, the smoke tests have failed, it seems, for unrelated things to the changes made, is it possible to repeat them?

theblixguy commented:
Debug failure is known (been failing for others as well) and is unrelated to your changes, so I wouldn’t worry about it.

@swift-ci please smoke test

minuscorp committed:
Fix python_support tests

minuscorp commented:
@swift-ci please smoke test cc @theblixguy

theblixguy commented:
Let’s just run python lint as that failed last time.

@swift-ci please python lint

minuscorp commented:
first time see that command

theblixguy commented:
Passed: https://ci.swift.org/view/Pull%20Request/job/swift-PR-python-lint/198/

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
